x/twap/listeners_test.go

```diff
@@ -92,49 +87,213 @@ func (s *TestSuite) TestAfterPoolCreatedHook() {
+			} else {
+				// height should not have changed
+				s.Require().Equal(twapAfterPoolCreation.Height, twapAfterBlock1.Height)
+				// twap time should be same as previous blocktime
+				s.Require().Equal(twapAfterPoolCreation.Time, baseTime)
+
+				// accumulators should not have increased, as they are going through the first epoch
+				s.Require().Equal(sdk.ZeroDec(), twapAfterBlock1.P0ArithmeticTwapAccumulator)
+				s.Require().Equal(sdk.ZeroDec(), twapAfterBlock1.P1ArithmeticTwapAccumulator)
+			}
+
+			// check if spot price has been correctly updated in twap record
+			asset0sp, err := s.App.GAMMKeeper.CalculateSpotPrice(s.Ctx, poolId, twapAfterBlock1.Asset0Denom, twapAfterBlock1.Asset1Denom)
+			s.Require().NoError(err)
+			s.Require().Equal(asset0sp, twapAfterBlock1.P0LastSpotPrice)
+
+			// run basic swap on block two for price change
+			if tc.block2Swap {
+				s.RunBasicSwap(poolId)
+			}
 
+			s.EndBlock()
+			s.Commit()
+
+			// Third block
+			twapAfterBlock2, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), denom0, denom1)
+			s.Require().NoError(err)
+
+			// if no swap happened in block 3, twap record should be same with block 2
+			if !tc.block2Swap {
+				s.Require().Equal(twapAfterBlock1, twapAfterBlock2)
+			} else {
+				s.Require().Equal(secondBlockTime, twapAfterBlock2.Time)
+
+				// check accumulators incremented - we test details of correct increment in logic
+				s.Require().True(twapAfterBlock2.P0ArithmeticTwapAccumulator.GT(twapAfterBlock1.P0ArithmeticTwapAccumulator))
+				s.Require().True(twapAfterBlock2.P1ArithmeticTwapAccumulator.GT(twapAfterBlock1.P1ArithmeticTwapAccumulator))
+			}
+
+			// check if spot price has been correctly updated in twap record
+			asset0sp, err = s.App.GAMMKeeper.CalculateSpotPrice(s.Ctx, poolId, twapAfterBlock1.Asset0Denom, twapAfterBlock2.Asset1Denom)
+			s.Require().NoError(err)
+			s.Require().Equal(asset0sp, twapAfterBlock2.P0LastSpotPrice)
+		})
+	}
+}
+
+// TestAfterEpochEnd tests if records get succesfully deleted via `AfterEpochEnd` hook.
+// We test details of correct implementation of pruning method in store test.
+// TODO: improve test using `GetAllHistoricalTimeIndexedTWAPs` and `GetAllHistoricalPoolIndexedTWAPs`
+func (s *TestSuite) TestAfterEpochEnd() {
+	s.SetupTest()
+	s.Ctx = s.Ctx.WithBlockTime(baseTime)
 	poolId := s.PrepareBalancerPoolWithCoins(defaultUniV2Coins...)
-	expectedHistoricalTwap, err := types.NewTwapRecord(s.App.GAMMKeeper, s.Ctx, poolId, denom0, denom1)
+	twapBeforeEpoch, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), denom0, denom1)
 	s.Require().NoError(err)
 
-	s.EndBlock()
-	s.Commit() // clear transient store
-	// Now on a clean state after a create pool
-	s.Require().Equal(baseTime.Add(time.Second), s.Ctx.BlockTime())
-	s.RunBasicSwap(poolId)
+	// epoch has passed
+	pruneEpochIdentifier := s.App.TwapKeeper.PruneEpochIdentifier(s.Ctx)
```

Comment:
errors here, errors bc params for twap is not registered

Reply:
Working on it: https://github.com/osmosis-labs/osmosis/pull/2451

Reply:
Ty for the fix!